Get timezone from menu button

KaterinaSolovyeva opened the pull request:
## Ticket - [Установка дефолтных кнопок для телеграма](https://www.notion.so/7d6d35f407814742a0cebea1e480bd4b?p=9f3edd440b0a4ffca5d77f5e695a2870)
добавлена кнопка Меню, из которой можно перейти в список доступных запросов либо установить часовой пояс

## Ticket - [Установка часового пояса + диалоги бота к нему](https://www.notion.so/7d6d35f407814742a0cebea1e480bd4b?p=599621be7b1f47e19470172398a5db7e)

При запросе на присвоение нового часового пояса появляются две кнопки объекта ReplyKeyboard:
'Отправить свою геолокацию для автоматической настройки часового пояса' - получает гелолокацию, по ней определяет таймзону. Передает по апи
'Напишу свою таймзону сам' - распознает введенный текст. Проверяет, существует ли такая таймзона. Если да, то передает её через апи. Если нет, то проверяет, не является ли введенный текст тем, на который отвечать не нужно. Просит пользователя ввести другую таймзону или выслать геолокацию.
После получения таймзоны кнопки ReplyKeyboard внизу исчезают

KaterinaSolovyeva committed:
Get timezone from menu button

KaterinaSolovyeva committed:
Get timezone from menu button

KaterinaSolovyeva committed:
Get timezone from menu button

MatMerd commented on requirements.txt:
```diff
@@ -1,42 +1,46 @@
-anyio==3.6.1
```
А как так получилось, что requirements перезаписался?

KaterinaSolovyeva replied:
Для определения таймзоны по локации я установила pip install timezonefinder[numba], он устанавливает дополнительные зависимости. Чтобы ничего не потерять, я добавила их все к requirements
Почему весь файл перезаписался не знаю..

MatMerd replied:
Вот тут сразу оговоримся. Давай не будем тянуть в проект numba. Эта зависимость не самая простая и требует бо́льших знаний о её использовании, чем у нас есть.

veselij replied:
я у себя пробовал когда запустить, у меня не встало, пришлось руками ставить и встала numpy на m1

MatMerd commented on src/bot.py:
```diff
@@ -1,66 +1,71 @@
-from telegram import Update
```
Опять странно. Перезаписался целый файл, а должна была поменяться одна строка, насколько я вижу

KaterinaSolovyeva replied:
не знаю, к сожалению
Мои шаги:
git checkout -b feature/menu_button_timezone
git add --all
git commit -m "Get timezone from menu button"
git push --set-upstream origin feature/menu_button_timezone

KaterinaSolovyeva replied:
после команды комита показали ошибки пре-комита, повторила add и commit перед пушем.
Что я сделала не так?

MatMerd replied:
Ощущение, что во время разработки, ты случайно перезаписала файл в другой. Или неудачно сделала merge

AntonZelinsky replied:
нет, не в другой, там по другому отображается название файла. Попробуй смержить к себе ветку девелоп, а потом сделай ребейз к ветке девелоп

MatMerd replied:
Антон наверное хотел написать сделать pull ветки девелоп, а потом сделать ребейз из своей ветки

KaterinaSolovyeva replied:
поняла почему, кажется. Я неправильно мержу вручную. Но не знаю как правильно. Помогите!
после пула некоторые файлы имеют разногласия со строками <<<<<<<старый код======<<<<<<<'второй вариант кода===== 
Что я делаю: удаляю в файле часть кода, которая не рабочая, оставляю правильную. Когда сохраняю, то весь код записывается под моим именем, даже если я выбрала чужой код
Подскажите, пожалуйста, как правильно делать?

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+
+from service.api_client_real import RealAPIService
+
+RHETORICAL_TEXT = ("Напишу свою таймзону сам", "Отправить свою геолокацию для автоматической настройки часового пояса")
```
Мне кажется, это должны быть две разные константы, а не кортеж

KaterinaSolovyeva replied:
в кортеже текст, на который бот не должен реагировать как на вариант часового пояса от пользователя

AntonZelinsky replied:
Да и вообще, помещай текст внутрь кнопки, пока не нужно выносить во вне

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+    )
+
+
+async def get_timezone_from_location(update: Update, context: CallbackContext):
```
Это получается очень большая и громоздкая функция. Её можно разбить на три части, как минимум:
1. Основная, которая отвечает за отправку таймзоны в апи фронта
2. Функция, которая определяет тайм зону из местоположения
3. Функция, которая определяет таймзону из текста

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+            api = RealAPIService()
+            await api.set_user_timezone(telegram_id=update.effective_chat.id, user_timezone=user_timezone)
+        except zoneinfo.ZoneInfoNotFoundError:
+            if text not in RHETORICAL_TEXT and set("/").isdisjoint(text):
```
Не понимаю первую часть условия. Как текст может входить в этот кортеж?

KaterinaSolovyeva replied:
когда пользователь нажимает на вторую кнопку "Напишу свою таймзону сам", то бот читает ее как ответ и пишет: "Часового пояса "Напишу свою таймзону сам" нет. Попробуйте написать другой или передайте геолокацию для его автоматического определения". А нужно, чтобы просто пропустил эту фразу и мы могли ввести часовой пояс руками

MatMerd replied:
А можно же сделать так, что после "Напишу свою таймзону сам, пользователю отправляется сообщение "Введите таймзону в формате <format>" и только после этого ожидать ввода таймзоны?

KaterinaSolovyeva replied:
А что если так: "Введите таймзону. Например: Europe/Moscow" и к этой строке инлайн кнопка: 
 "посмотреть список всех таймзон в Википедии" - [ссылка на википедию](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)

MatMerd replied:
А сама бы ты как отнеслась к такому сервису? Я бы , например, очень негодовал. Особенно когда ты врач и хочешь войти в систему по рабочим вопросам.

AntonZelinsky replied:
Блин, я понял, Катя сделала KeyboardButton, а это такие серые кнопки, при нажатии на которую отправляется сообщение от имени пользователя. Не, мы такие кнопки использовать точно не будем, мы используем только инлайновые кнопки. В примерах с конверсейшенами такие тоже есть

MatMerd replied:
Точно!
Спасибо, Антон

KaterinaSolovyeva replied:
Именно через серые кнопки можно получить геолокацию пользователя. У пользователя всплывает окошко с запросом к нему на разрешение получения его геолокации. Поэтому я использовала именно их для этого.
[KeyboardButton('Отправить свою геолокацию для автоматической настройки часового пояса', **request_location=True**)]
Почему мы не будем использовать эти кнопки? Можно ли в данном случае сделать исключение?

AntonZelinsky replied:
Да, ты права. тут будет исключение, потому что по другому не получится запросить локацию.

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+    elif update.message.text:
+        text = str(update.message.text)
+        try:
+            zoneinfo.ZoneInfo(text)
```
В данном варианте получается, что человеку нужно четко знать, какая у него таймзона и как она пишется. Например, я понятия не имею, что напишу, если окажусь в Бразилии. Нужен более элегантный механизм получения часового пояса. Можно подумать в сторону того, что человек может вводить текущее время, мы находим разницу по часам с москвой (или с гринвичем) и от этого уже считаем часовой пояс. Формат времени у всех одинаковый

KaterinaSolovyeva replied:
по условию Александра Минаева: "У пользователя должен быть выбор или расшарить геолокацию или ввести руками таймзону"
Поэтому первая кнопка - это тот самый элегантный способ)) получаем геолокацию и по ней считается таймзона. 
Уточни, нужно ли переделывать второй вариант все-таки?

MatMerd replied:
Давай позовём сюда Сашу и Антона. Я думаю, что именно этот момент тебе сейчас и не нужно будет переделывать
@AntonZelinsky @veselij

veselij replied:
мы изначально отказались почему то от текущего времени, всякие пограничные состояния наверное редко, но все же могут быть, типа реально время 17:55, а человеку лень писать и он такой хоп 18 отправил. и это и парсить сложно и на час промахнуться можно. я думал таймзону отправлять в формате UTC+3 UTC+4 и т.д. и это в примере написать можно и распарсить легко

KaterinaSolovyeva replied:
о, теперь поняла! сделаю )

AntonZelinsky replied:
Можно даже просто `+3`, `-5`, `+9`. Так же, добавь кнопку поделиться геолокацией, она была в примерах к нашей библиотеке (ищи примеры с конверсейшенами) и кнопку `+3`.
Сторонние библиотеки точно не будем тянуть.

KaterinaSolovyeva replied:
просто цифрами, мне кажется, непонятно. Будут разницу с Москвой писать. сделала как: UTC+3
кнопку с геолокацией сделала через автоматический запрос к пользователю:
[KeyboardButton('Отправить свою геолокацию для автоматической настройки часового пояса', **request_location=True**)]
"она была в примерах к нашей библиотеке (ищи примеры с конверсейшенами)" - прошу прощения, эту часть вообще не поняла. Это что и где смотреть?

MatMerd replied:
Это смотреть в telegram-bot-api. Там есть вики и примеры разного функционала.

veselij replied:
я тоже не нашел, если мы хотим показать пользователю кнопку запросить локацию - то только так. онлайн кнопками такое не сделать. можно попросто написать сообщение "расширьте нам свою геолокацию через стандартную клавиатуру" и тогда мы его сможем отработать. Но тут сомнения все ли поймут, что мы просим. можно конечно в сообщении прям инструкцию написать или скрин отправить типа куда жать

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+            text=f'Установлен часовой пояс "{user_timezone}"',
+            reply_markup=ReplyKeyboardRemove(),
+        )
+        api = RealAPIService()
```
У нас есть реализация фейкового сервиса. Использование просто RealAPIService это черевато, пока он не готов. Посмотри как использовать разные реализации сервиса в src/service/__init__.py

KaterinaSolovyeva replied:
да, тестила на нем, думала для проверки отправить все-таки реальный. Ок, поправлю

AntonZelinsky commented on src/get_timezone.py:
```diff
@@ -0,0 +1,62 @@
+
+
+get_timezone_command_handler = CommandHandler("get_timezone", get_timezone)
+get_timezone_from_location_handler = MessageHandler(filters.TEXT | filters.LOCATION, get_timezone_from_location)
```
И для обработки трёх вариантов дейстивия пользователей (нажатие инлайновой кнопки, ввода текста и отправки локации) нужно использовать три разные хэндлена! Тогда везде будет очень простой код

KaterinaSolovyeva committed:
Merging

KaterinaSolovyeva committed:
Merging

KaterinaSolovyeva committed:
Merging

KaterinaSolovyeva committed:
Changed structure of get_timezone

1kovalevskiy committed:
init auth_procedure

AigulParamonova committed:
fix command constants and state names

AigulParamonova committed:
fixed a bug

veselij committed:
Merge pull request #27 from Studio-Yandex-Practicum/feature/correct_command_constants

fix command constants and state names

1kovalevskiy committed:
create auth_telegram_user and example

1kovalevskiy committed:
refactored and add docstring

1kovalevskiy committed:
rename config constants and add comments

1kovalevskiy committed:
fix imports in api_client and add TODO + black edit codestile

1kovalevskiy committed:
embedded auth to start_conversation

1kovalevskiy committed:
create docstring

1kovalevskiy committed:
deleted unused files

1kovalevskiy committed:
merge develop into feature/auth_procedure

AntonZelinsky committed:
Merge pull request #28 from Studio-Yandex-Practicum/feature/auth_procedure

Add auth function with use case example

zhss1983 committed:
Changed the names of constants and their values where necessary.

agamova committed:
Dockerfile support added

zhss1983 committed:
Combined the bot_menu and start_conversation files. Replaced the "start" command with a constant that was previously created, but was not used in the project.

zhss1983 committed:
Highlighted unused constants.

zhss1983 committed:
I moved the constants from the project to a separate folder. I didn't make any changes, the status can be viewed by the previous commit.

zhss1983 committed:
Deleted unused constants

zhss1983 committed:
Moved callbacks to callback_data file

zhss1983 committed:
Add cansel command to constants

zhss1983 committed:
Restored constants folder and moved URL_SERVICE_RULES to enviroments

zhss1983 committed:
Removed cansel_comand and it's methods.

zhss1983 committed:
Removed cansel_comand and it's methods. Fix PEP8 error

zhss1983 committed:
Rename callback constants to command ones.

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Изменил имена констант на CALLBACK_<имя метода который он вызывает>_COMMAND для того чтобы сразу становилось ясно для чего нужна та или иная константа.
Перенёс запуск меню в start_conversation по команде /menu.

zhss1983 committed:
I changed the names of the constants to CALLBACK_<the name of the method that it calls>_COMMAND so that it immediately becomes clear what this or that constant is needed for.
Moved the menu launch to start_conversation by the /menu command.
I saved callback_menu_handler in case it is imported in the future, the code is not mine, I do not know how it will be reused.

agamova committed:
start method changed

KaterinaSolovyeva committed:
Fixed conflicts in PR

KaterinaSolovyeva committed:
Fixed conflicts in PR

KaterinaSolovyeva committed:
Fixed conflicts in PR

AntonZelinsky committed:
Merge pull request #31 from Studio-Yandex-Practicum/refactoring

Refactoring

agamova committed:
gunicorn removed

agamova committed:
added comment to Dockerfile

AntonZelinsky committed:
Merge pull request #30 from Studio-Yandex-Practicum/feature/Dockerfile

Dockerfile support added

KaterinaSolovyeva committed:
Fixed conflicts in PR
